README.md
# ASN1bean
ASN1bean (previously known as jASN1) is a Java ASN.1 BER and DER encoding/decoding library

For detailed information on ASN1bean visit https://www.beanit.com/asn1/.


doc/CHANGELOG.txt
v1.14.0 30-Jun-2023
--------------------

- add support for decimal REAL decoding, thanks to bibmaster and karim

- fix Windows run scripts for newer Java versions


v1.13.0 30-Jul-2021
--------------------

- added support for EXTENSIBILITY APPLIED when decoding sequences:
  accessing and setting extension bytes is possible using "-e" flag
  when compiling ASN1 files, thanks for testing to Calum MacLean

- fixed decoding empty ANY value, thanks to Steffen Conrad

- fixed decoding ANY containing element with indefinite length, thanks
  to Armin Lunkeit for pointing to the problem

- add support for decoding constructed OCTET STRINGs


v1.12.0 20-Jun-2020
--------------------

- renamed jASN1 to ASN1bean

- Added full support for indefinite length decoding. The compiler's
  -il flag to enable indefinite length decoding was removed because it
  is now always enabled.

- added support for white space in compiler's path parameters

- added support for ASN1 modules that start with a digit

- multiple minor clean ups in generated code


v1.11.3 14-Mar-2020
--------------------

- switched from Java 7 to Java 8

- Improved parsing of scientific number like 7.0E-2 and single value
  constraint for octet strings like OCTET STRING (SIZE( 1)) ('55'H)

- certain self reference like A ::= A; now cause an error message

- minor fix in run script to allow for spaces in parent directories


v1.11.2 06-Jun-2019
--------------------

- added compile support for various forms of INTEGER and REAL value
  constraints

- significant performance improvements to
  ReverseByteArrayOutputStream.write(byte[]) and
  BerLength.encodeLength

Thanks to the European Space Agency for these contributions.


v1.11.1 03-May-2019
--------------------

- removed use of JAXB function to make jASN1 compatible with Java 11

- insertion of jASN1 version in generated class headers can now be
  disabled with -dv flag

- added support for information object classes defined in external
  modules

- fixed name clash issue if a sequence, set or choice has an element
  with the same name as the sequence, set or choice

- fixed several minor inefficiencies such as unnecessary autoboxing


v1.11.0 15-Mar-2019
--------------------

- changed license to Apache 2.0 license

- changed package names from org.openmuc.* to com.beanit.*

- run scripts are now found in the bin/ folder


v1.10.0 26-Jun-2018
-------------------

- by default compiler now creates classes that implement the general
  BerType interface which contains the encode and decode functions,
  behavior can be disabled with compilers -di flag

- made asCalendar() and asDate() methods of BerGeneralizedTime class
  public

- open types in information object classes may now be OPTIONAL

- fixed support of ellipsis ('...') in the beginning of sequences


v1.9.0 13-Dec-2017
------------------

- encode functions now take general OutputStream interface as
  parameter instead of BerByteArrayOutputStream so that alternative
  reverse output streams could be developed and used. thanks to önder
  sezgin

- renamed BerByteArrayOutputStream to ReverseByteArrayOutputStream

- the BerBitString value can now be set and retrieved as array of
  booleans


v1.8.2 28-Aug-2017
------------------

- fixed decoding certain subIDs of object identifiers, thanks to Calum
  MacLean


v1.8.1 19-Jul-2017
------------------

- changed license to Mozilla Public License (MPL) v2.0

- improved BerGeneralizedTime.java thanks to Andrea Lo Pumo

- fixed decoding VisibleStrings to use readFully so that no exception
  is thrown while if read return less bytes, thanks to Jonathan
  Huthmann and Andrea Lo Pumo


v1.8.0 20-Apr-2017
------------------

- generated classes only import types imported by its module, not all
  available modules. This prevents name clashes.

- added support for parametrized types

- ANY types are now decoded and encoded differently so that their
  contents can be easily decoded by another type.

- removed support for ANY_NODECODE

- improved behaviour of compiler interface:

  * package folders no longer need to be part of the base directory
    specified.

  * files are separated by white space instead of ','

- added support for Information Object Classes

- added support for ENUMERATED lists without value and INTEGERS with
  both a value list and a constraint

- OID values are now parsed and stored in special class OidValues.java

- correctly fixed name clash of generated classes from the following
  structure SEQUENCE { test SEQUENCE OF Test }

- packages, fields and methods that clash with reserved Java key words
  are renamed (thanks to Telespazio Vega)

- added support for EMBEDDED PDV type (thanks to Telespanzio Vega)

- toString() methods of CHOICE, SEQUENCE/SET and SEQUENCE/SET OF use
  indentation to format returned string.

- added support for TIME, TIME-OF-DAY, DATE, DATE-TIME, DURATION types

- added asCalendar() and asDate() functions to BerUtcTime and
  BerGeneralizedTime thanks to Andrea Lo Pumo


v1.7.1 05-Feb-2017
------------------

- removed dependency on javax.xml.bind.DatatypeConverter so it works
  on Android again. Thanks to Karsten Ohme

- fixed coding ENUMERATED bug, introduced in release 1.7.0, thanks to
  David Jelenc

- fixed two more bugs when encoding and decoding REALs, thanks to htw
  saar


v1.7.0 27-Jan-2017
------------------

- fixed bug that the tag default of one module would override that of
  another module.

- added support for explicitly tagged SEQUENCE, SET, SEQUENCE/SET OF,
  CHOICE and ANY

- JAXB syntax mode of jasn-compiler is now default. Old syntax can be
  enabled using the legacy (-l) flag.

- numerous minor improvements to jasn1-compiler to create more
  comprehensible Java classes.

- BerInteger is now based on BigInteger instead of long. Thus it can
  code integer values larger than 8 bytes.

- fixed name clash when a sequence element of type SEQUENCE OF has the
  same name as the SEQUENCE OF type, e.g. SEQUENCE { test SEQUENCE OF

projects/asn1bean/src/main/java/com/beanit/asn1bean/ber/BerTag.java
/*
 * Copyright 2012 The ASN1bean Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.beanit.asn1bean.ber;

import com.beanit.asn1bean.util.HexString;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;

public class BerTag implements Serializable {

  public static final int UNIVERSAL_CLASS = 0x00;
  public static final int APPLICATION_CLASS = 0x40;
  public static final int CONTEXT_CLASS = 0x80;
  public static final int PRIVATE_CLASS = 0xc0;
  public static final int PRIMITIVE = 0x00;
  public static final int CONSTRUCTED = 0x20;
  public static final int BOOLEAN_TAG = 1;
  public static final int INTEGER_TAG = 2;
  public static final int BIT_STRING_TAG = 3;
  public static final int OCTET_STRING_TAG = 4;
  public static final int NULL_TAG = 5;
  public static final int OBJECT_IDENTIFIER_TAG = 6;
  public static final int OBJECT_DESCRIPTOR_TAG = 7;
  public static final int REAL_TAG = 9;
  public static final int ENUMERATED_TAG = 10;
  public static final int UTF8_STRING_TAG = 12;
  public static final int TIME_TAG = 14;
  public static final int SEQUENCE_TAG = 16;
  public static final int SET_TAG = 17;
  public static final int NUMERIC_STRING_TAG = 18;
  public static final int PRINTABLE_STRING_TAG = 19;
  public static final int TELETEX_STRING_TAG = 20;
  public static final int VIDEOTEX_STRING_TAG = 21;
  public static final int IA5_STRING_TAG = 22;
  public static final int UTC_TIME_TAG = 23;
  public static final int GENERALIZED_TIME_TAG = 24;
  public static final int GRAPHIC_STRING_TAG = 25;
  public static final int VISIBLE_STRING_TAG = 26;
  public static final int GENERAL_STRING_TAG = 27;
  public static final int UNIVERSAL_STRING_TAG = 28;
  public static final int BMP_STRING_TAG = 30;
  public static final int DATE_TAG = 31;
  public static final int TIME_OF_DAY_TAG = 32;
  public static final int DATE_TIME_TAG = 33;
  public static final int DURATION_TAG = 34;
  public static final BerTag SEQUENCE = new BerTag(UNIVERSAL_CLASS, CONSTRUCTED, SEQUENCE_TAG);
  public static final BerTag SET = new BerTag(UNIVERSAL_CLASS, CONSTRUCTED, SET_TAG);
  private static final long serialVersionUID = 1L;
  public byte[] tagBytes = null;
  public int tagClass;
  public int primitive;
  public int tagNumber;

  public BerTag(int identifierClass, int primitive, int tagNumber) {
    this.tagClass = identifierClass;
    this.primitive = primitive;
    this.tagNumber = tagNumber;
    code();
  }

  public BerTag() {}

  private void code() {
    if (tagNumber < 31) {
      tagBytes = new byte[1];
      tagBytes[0] = (byte) (tagClass | primitive | tagNumber);
    } else {
      int tagLength = 1;
      while (tagNumber > (Math.pow(2, (7 * tagLength)) - 1)) {
        tagLength++;
      }

      tagBytes = new byte[1 + tagLength];
      tagBytes[0] = (byte) (tagClass | primitive | 31);

      for (int j = 1; j <= (tagLength - 1); j++) {
        tagBytes[j] = (byte) (((tagNumber >> (7 * (tagLength - j))) & 0xff) | 0x80);
      }

      tagBytes[tagLength] = (byte) (tagNumber & 0x7f);
    }
  }

  public int encode(OutputStream reverseOS) throws IOException {
    if (tagBytes == null) {
      code();
    }
    for (int i = (tagBytes.length - 1); i >= 0; i--) {
      reverseOS.write(tagBytes[i]);
    }
    return tagBytes.length;
  }

  public int encodeForwards(OutputStream os) throws IOException {
    if (tagBytes == null) {
      code();
    }
    for (int i = 0; i < tagBytes.length; i++) {
      os.write(tagBytes[i]);
    }
    return tagBytes.length;
  }

  public int decode(InputStream is) throws IOException {
    int nextByte = is.read();
    if (nextByte == -1) {
      throw new EOFException("Unexpected end of input stream.");
    }

    tagClass = nextByte & 0xC0;
    primitive = nextByte & 0x20;
    tagNumber = nextByte & 0x1f;

    int codeLength = 1;

    if (tagNumber == 0x1f) {
      tagNumber = 0;
      int numTagBytes = 0;

      do {
        nextByte = is.read();
        if (nextByte == -1) {
          throw new EOFException("Unexpected end of input stream.");
        }

        codeLength++;
        if (numTagBytes >= 6) {
          throw new IOException("Tag is too large.");
        }
        tagNumber = tagNumber << 7;
        tagNumber |= (nextByte & 0x7f);
        numTagBytes++;
      } while ((nextByte & 0x80) != 0);
    }
    tagBytes = null;

    return codeLength;
  }

  public int decode(InputStream is, OutputStream os) throws IOException {
    int nextByte = is.read();
    if (nextByte == -1) {
      throw new EOFException("Unexpected end of input stream.");
    }
    os.write(nextByte);

    tagClass = nextByte & 0xC0;
    primitive = nextByte & 0x20;
    tagNumber = nextByte & 0x1f;

    int codeLength = 1;

    if (tagNumber == 0x1f) {
      tagNumber = 0;
      int numTagBytes = 0;

      do {
        nextByte = is.read();
        if (nextByte == -1) {
          throw new EOFException("Unexpected end of input stream.");
        }
        os.write(nextByte);

        codeLength++;
        if (numTagBytes >= 6) {
          throw new IOException("Tag is too large.");
        }
        tagNumber = tagNumber << 7;
        tagNumber |= (nextByte & 0x7f);
        numTagBytes++;
      } while ((nextByte & 0x80) != 0);
    }
    tagBytes = null;

    return codeLength;
  }

  /**
   * Decodes the Identifier from the ByteArrayInputStream and throws an Exception if it is not equal
   * to itself. Returns the number of bytes read from the InputStream.
   *
   * @param is the input stream to read the identifier from.
   * @return the length of the identifier read.
   * @throws IOException if an exception occurs reading the identifier from the stream.
   */
  public int decodeAndCheck(InputStream is) throws IOException {
    if (tagBytes == null) {
      code();
    }
    for (byte identifierByte : tagBytes) {
      int nextByte = is.read();
      if (nextByte == -1) {
        throw new EOFException("Unexpected end of input stream.");
      }

      if (nextByte != (identifierByte & 0xff)) {
        throw new IOException(
            "Identifier does not match, expected: 0x"
                + HexString.fromByte(identifierByte)
                + ", received: 0x"
                + HexString.fromByte((byte) nextByte));
      }
    }
    return tagBytes.length;
  }

  public boolean equals(int identifierClass, int primitive, int tagNumber) {
    return (this.tagNumber == tagNumber
        && this.tagClass == identifierClass
        && this.primitive == primitive);
  }

  @Override
  public boolean equals(Object obj) {
    if (!(obj instanceof BerTag)) {
      return false;
    }
    if (obj == this) {
      return true;
    }
    BerTag berIdentifier = (BerTag) obj;
    return (tagNumber == berIdentifier.tagNumber
        && tagClass == berIdentifier.tagClass
        && primitive == berIdentifier.primitive);
  }

  @Override
  public int hashCode() {
    int hash = 17;
    hash = hash * 31 + tagNumber;
    hash = hash * 31 + tagClass;
    hash = hash * 31 + primitive;
    return hash;
  }

  @Override
  public String toString() {
    return "identifier class: "
        + tagClass
        + ", primitive: "
        + primitive
        + ", tag number: "
        + tagNumber;
  }
}
